xla/service/gpu/priority_fusion.cc

```diff
@@ -247,13 +252,44 @@ class GpuPriorityFusionQueue {
     return !current_consumers_.empty();
   }
 
+  void UpdatePerformanceModelCache(HloInstruction* producer) {
+    if (producer->opcode() == HloOpcode::kBitcast ||
+        producer->opcode() == HloOpcode::kConstant || !IsFusible(*producer)) {
+      return;
+    }
+
+    for (auto consumer : producer->users()) {
+      if (!IsFusible(*consumer)) return;
+    }
+
+    auto config = GpuPerformanceModelOptions::PriorityFusion(
+        &fusion_analysis_cache_, &gpu_performance_model_cache_);
+
+    if (!gpu_performance_model_cache_.Get(*producer)) {
+      auto runtime_data = GpuPerformanceModel::EstimateRunTimeForInstruction(
+          producer, *device_info_, &cost_analysis_, config);
+      gpu_performance_model_cache_.Set(*producer, runtime_data);
+    }
+
+    for (auto consumer : producer->users()) {
```

Comment:
It's not necessary to update costs for consumers here. If everything is done right, the function will be also called for all fusible instructions in the computation anyway.

Reply:
That's right. Except there is a case where bitcast can still have consumers after fusing as producer. Therefore, after cache invalidate, it can still be in the graph causing cache miss.

Reply:
Is this because of `producer->opcode() == HloOpcode::kBitcast` check above? I think we can remove that check, it's not necessary.

Reply:
https://github.com/openxla/xla/blob/0a4d157ba44b378221a915965036c0aff0120a70/xla/service/gpu/priority_fusion.cc#L241 here, bitcast can still have consumers after fusing with all its consumers so after cache invalidation, you wont find bitcast in the cache but it can still be required. I changed the logic to only invalidate if the producer has 0 consumers. See my latest commit.